lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,126 @@
+//===-- lib/semantics/check-select-stmt.cpp -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/Common/Fortran.h"
+#include "flang/Common/idioms.h"
+#include "flang/Parser/message.h"
+#include "flang/Parser/tools.h"
+#include "flang/Semantics/tools.h"
+#include <list>
+#include <optional>
+#include <set>
+#include <tuple>
+#include <variant>
+
+namespace Fortran::semantics {
+
+void SelectConstructChecker::Leave(
+    const parser::SelectRankConstruct &selectRankConstruct) {
+  const auto &selectRankStmt{
+      std::get<parser::Statement<parser::SelectRankStmt>>(
+          selectRankConstruct.t)};
+  const auto &selectRankStmtSel{
+      std::get<parser::Selector>(selectRankStmt.statement.t)};
+
+  // R1149 select-rank-stmt checks
+  const Symbol *saveSelSymbol{nullptr};
+  if (const auto resolvedSel{ResolveSelector(selectRankStmtSel)}) {
+    if (const Symbol * sel{evaluate::UnwrapWholeSymbolDataRef(*resolvedSel)}) {
+      if (!IsAssumedRankArray(*sel)) {  // C1150
+        context_.Say(parser::FindSourceLocation(selectRankStmtSel),
+            "Selector '%s' is not an assumed-rank array variable"_err_en_US,
+            sel->name().ToString());
+      } else {
+        saveSelSymbol = sel;
+      }
+    } else {
+      context_.Say(parser::FindSourceLocation(selectRankStmtSel),
+          "Selector '%s' is not an assumed-rank array variable"_err_en_US,
```

Comment:
Is this the right error message for selectors that are not whole variables?

Reply:
If I remember, I saw somwthing similar on gfortran front, not sure though.
But can you suggest a better one?